libdrgn/debug_info.c

```diff
@@ -1201,7 +1212,52 @@ drgn_lazy_type_from_dwarf(struct drgn_debug_info *dbinfo, Dwarf_Die *parent_die,
 	thunk->thunk.free_fn = drgn_type_from_dwarf_thunk_free_fn;
 	thunk->die = type_die;
 	thunk->can_be_incomplete_array = can_be_incomplete_array;
-	drgn_lazy_type_init_thunk(ret, &thunk->thunk);
+	thunk->bias = bias;
+	drgn_lazy_parameter_init_type_thunk(ret, &thunk->thunk);
+	return NULL;
+}
+
+static struct drgn_error *
+drgn_object_from_dwarf_thunk_evaluate_fn(struct drgn_object_thunk *thunk,
+					 struct drgn_object *ret);
+static void drgn_object_from_dwarf_thunk_free_fn(struct drgn_object_thunk *thunk)
+{
+	free(container_of(thunk, struct drgn_object_from_dwarf_thunk, thunk));
+}
+
+
+static struct drgn_error *
+drgn_lazy_object_from_dwarf(struct drgn_debug_info *dbinfo,
+			    Dwarf_Die *die,
+			    uint64_t bias,
+			    const char* name,
+			    struct drgn_lazy_parameter *ret)
+{
+	struct drgn_object_from_dwarf_thunk *thunk = malloc(sizeof(*thunk));
+	if (!thunk)
+		return &drgn_enomem;
+
+	thunk->thunk.prog = dbinfo->prog;
+	thunk->thunk.evaluate_fn = drgn_object_from_dwarf_thunk_evaluate_fn;
+	thunk->thunk.free_fn = drgn_object_from_dwarf_thunk_free_fn;
+	thunk->name = name;
+
+	bool decl;
+	if (!dwarf_flag(die, DW_AT_declaration, &decl) && decl) {
+		Dwfl_Module *module;
+		Dwarf_Off offset;
+		if (drgn_dwarf_index_find_definition(&dbinfo->dindex, (uintptr_t)die->addr, &module, &offset)) {
```

Comment:
Note to self - if we aren't able to find a declaration here, the object returned is pretty odd and can probably cause a crash if mishandled. Not sure if we should worry about that at the moment - but I have been able to generate class functions with no declarations (usually if the function was not used or defined)